### PR TITLE
(CFACT-83) Implement a public API for libfacter.

### PR DESCRIPTION
**Pull request by peterhuene**
Implementing a public API for libfacter.

This separates the include headers into a public API that is enough for third parties to define their own facts and use the collection, but hides the "internal" implementation details.

On Linux and Windows, we now export only the public interface, so libfacter should be smaller as a result.

### CODE REVIEW

**Comment by peterhuene** [comment]
I tried for far too long to get this to work properly on OSX.  I can't hide third-party exports as easily and even with an "unexported list file", I couldn't get Boost Log to actually output any messages.  For now I'd rather get these changes merged in and figure out hiding symbols on OSX some time in the future.

**Comment by branan** [comment]
This all looks pretty sane to me.

I think we need to revisit logging at some point, but I have no objections to the wrapper in the short term.

**Comment by peterhuene** [comment]
Yeah, not a huge fan of duplicating the logging interface, but I think it makes sense to have consumers of libfacter use a consistent namespace layout, provided the logging implementation is still all in leatherman.

**Comment by branan** [comment]
Yeah. This gets weird when we have, say, a consumer of both libfacter and libcthun. I've been mulling over how to handle this and will have an email to client@ sometime soon with a few thoughts and proposals.

Logging is hard, let's go shopping.

**Comment by peterhuene** [comment]
Reviewers: please hold off on merge until 0.4.0 is tagged.

File: exe/cfacter.cc [hunk 1]
```diff
@@ -1,8 +1,7 @@
 #include <facter/version.h>
-#include <leatherman/logging/logging.hpp>
+#include <facter/logging/logging.hpp>
```
Comment (MikaelSmith) [inline exe/cfacter.cc]: I'm not clear why we can't use leatherman's logging here.

Comment (MikaelSmith) [inline exe/cfacter.cc]: Oh, you're specifically exporting logging. I'd rather that were locked down more and not duplicated, but we can handle that separately.

Comment (peterhuene) [inline exe/cfacter.cc]: I wanted the cfacter executable to be an exemplar of consuming libfacter.  That means not relying on things we don't ship as part of libfacter, like leatherman.  cfacter also shouldn't care about the logging namespace, etc, and always rely on the implementation exported from libfacter.

Comment (MikaelSmith) [inline exe/cfacter.cc]: Ah... I probably would've leaned the other way and said if you want logging in your consumer, use leatherman.

Comment (peterhuene) [inline exe/cfacter.cc]: I view leatherman as something we link into PL libraries and not really something third parties make use of, at least that's my feelings towards it.

Comment (MikaelSmith) [inline exe/cfacter.cc]: As Branan said, we should revisit this later. I'll live with it for now.

**Comment by MikaelSmith** [comment]
Typo in the 2nd commit message.

**Comment by peterhuene** [comment]
Also looks like I need to amend those last two commits.  I did them from a vagrant VM.

**Comment by MikaelSmith** [comment]
Yup, `Vagrant User` looks kind of funny.

**Comment by peterhuene** [comment]
Corrected.

**Comment by MikaelSmith** [comment]
:+1:

**Pull request merged**